File: marshaller.py
def marshal(object):
    # Initialize an empty list to hold the string representations of the key/value pairs.
    string_array = []

    # Iterate over the key/value pairs in the dictionary.
    for key, value in object.items():
        # For each pair, convert the key and value to strings (if they aren't already),
        # join them with a '/', and append the result to 'string_array'.
        string_array.append(str(key) + '/' + str(value))
    
    # Join all the elements in 'string_array' with a '|', resulting in a single string
    # where each key/value pair is separated by '/', and each pair is delimited by '|'.
    string = '|'.join(string_array)

    # Encode the final string into bytes using the default encoding (utf-8) and return it.
    return string.encode('utf-8')

def unmarshal(received: bytes) -> object:
    # Decode the bytes argument 'received' into a UTF-8 string.
    received = received.decode('utf-8')

    # Split the decoded string by '|' to create a list, where each element
    # is expected to be a key/value pair joined by '/'.
    received_arr = received.split('|')

    # Initialize an empty dictionary to hold the key/value pairs.
    obj = {}

    # Iterate through each key/value pair in the list 'received_arr'.
    for pair in received_arr:
        # Split each pair by '/' to separate the key and value.
        key, val = pair.split('/')

        # Assign the value to its corresponding key in the 'obj' dictionary.
        obj[key] = val

    # Return the dictionary containing all the key/value pairs.
    return obj

File: Marshaller.java
// Marshaller.java
import java.io.*;
import java.nio.charset.StandardCharsets;
import java.util.Map;
import java.util.HashMap;

public class Marshaller {
    /**
     * Marshals a Map<String, String> object into a byte array.
     * The key-value pairs are encoded as "key/value" strings separated by "|" delimiter.
     *
     * @param object The Map<String, String> object to be marshaled.
     * @return The marshaled byte array representation of the object.
     */
    public static byte[] marshal(Map<String, String> object) {
        StringBuilder sb = new StringBuilder();
        boolean first = true; // Flag to check if it's the first iteration
        // Iterate over the entries in the Map
        for (Map.Entry<String, String> entry : object.entrySet()) {
            if (first) {
                first = false; // It's no longer the first iteration
            } else {
                sb.append('|'); // Append delimiter before each entry except the first
            }
            sb.append(entry.getKey()).append('/').append(entry.getValue()); // Append "key/value" string
        }
        // Convert the StringBuilder to a byte array using UTF-8 encoding
        return sb.toString().getBytes(StandardCharsets.UTF_8);
    }
    /**
     * Unmarshals a byte array into a Map<String, String> object.
     * The byte array is expected to be in the format "key/value|key/value|...",
     * where "|" is the delimiter separating key-value pairs.
     *
     * @param received The byte array to be unmarshaled.
     * @param offset   The offset in the byte array where the data starts.
     * @param length   The length of the data in the byte array.
     * @return The unmarshaled Map<String, String> object.
     */
    public static Map<String, String> unmarshal(byte[] received, int offset, int length) {
        Map<String, String> obj = new HashMap<>();
        String receivedString = new String(received, offset, length, StandardCharsets.UTF_8); // Convert byte array to string
        // Split the string into key-value pairs using the "|" delimiter
        String[] pairs = receivedString.split("\\|");
        for (String pair : pairs) {
            if (!pair.isEmpty()) {
                String[] keyValue = pair.split("/"); // Split the pair into key and value
                obj.put(keyValue[0], keyValue[1]); // Add the key-value pair to the Map
            }
        }
        return obj;
    }
}
